File: backend/app/schemas/intiaro.py
from __future__ import annotations
from typing import Optional
from pydantic import BaseModel


# --- ProductFeatures ---

class ProductFeaturesBase(BaseModel):
    web_gltf: bool = False
    web_usdz: bool = False
    web_360: bool = False
    app_ar: bool = False
    download3d: bool = False
    com: bool = False
    has_measurements: bool = False
    resolution_x: Optional[int] = None
    resolution_y: Optional[int] = None
    supported_formats: Optional[list] = None

class ProductFeaturesCreate(ProductFeaturesBase):
    pass

class ProductFeaturesOut(ProductFeaturesBase):
    id: int
    product_id: int
    model_config = {"from_attributes": True}


# --- RenderSettings ---

class RenderSettingsBase(BaseModel):
    rotate_size_x: Optional[int] = None
    rotate_size_y: Optional[int] = None
    tile_size_x: Optional[int] = None
    tile_size_y: Optional[int] = None
    zoom_size_x: Optional[int] = None
    zoom_size_y: Optional[int] = None
    shadow_enabled: bool = False

class RenderSettingsCreate(RenderSettingsBase):
    pass

class RenderSettingsOut(RenderSettingsBase):
    id: int
    product_id: int
    model_config = {"from_attributes": True}


# --- VariableGroup ---

class VariableGroupBase(BaseModel):
    name: str
    slug: Optional[str] = None
    icon: Optional[str] = None
    icon_selected: Optional[str] = None
    color: Optional[str] = None
    index: int = 0
    priority: int = 0
    description: Optional[str] = None

class VariableGroupCreate(VariableGroupBase):
    pass

class VariableGroupOut(VariableGroupBase):
    id: int
    product_id: int
    model_config = {"from_attributes": True}


# --- ChoiceGroup ---

class ChoiceGroupBase(BaseModel):
    name: str
    slug: Optional[str] = None
    index: int = 0

class ChoiceGroupCreate(ChoiceGroupBase):
    pass

class ChoiceGroupOut(ChoiceGroupBase):
    id: int
    product_id: int
    model_config = {"from_attributes": True}


# --- ProductPredicate ---

class ProductPredicateBase(BaseModel):
    predicate_key: str
    name: Optional[str] = None
    type: Optional[str] = None
    attribute: Optional[str] = None
    operator: Optional[str] = None
    compare_to: Optional[str] = None

class ProductPredicateCreate(ProductPredicateBase):
    pass

class ProductPredicateOut(ProductPredicateBase):
    id: int
    product_id: int
    model_config = {"from_attributes": True}


# --- ProductEvent ---

class ProductEventBase(BaseModel):
    trigger_variable: Optional[str] = None
    source_type: Optional[str] = "variable"
    source_variable: Optional[str] = None
    destinations: Optional[list] = None
    predicate_key: Optional[str] = None
    condition_attribute: Optional[str] = None
    condition_operator: Optional[str] = None
    condition_compare_to: Optional[str] = None

class ProductEventCreate(ProductEventBase):
    pass

class ProductEventOut(ProductEventBase):
    id: int
    product_id: int
    model_config = {"from_attributes": True}


# --- SectionalElement ---

class SectionalElementBase(BaseModel):
    element_id: Optional[int] = None
    name: Optional[str] = None
    display_name: Optional[str] = None
    file_id: Optional[str] = None
    default_variables: Optional[dict] = None
    includes: Optional[dict] = None
    positions: Optional[dict] = None

class SectionalElementCreate(SectionalElementBase):
    pass

class SectionalElementOut(SectionalElementBase):
    id: int
    product_id: int
    model_config = {"from_attributes": True}


# --- MenuSettings ---

class MenuSettingsBase(BaseModel):
    name: Optional[str] = None
    hidden_attributes: Optional[list] = None

class MenuSettingsCreate(MenuSettingsBase):
    pass

class MenuSettingsOut(MenuSettingsBase):
    id: int
    product_id: int
    model_config = {"from_attributes": True}


# --- AttributeMapping ---

class AttributeMappingBase(BaseModel):
    source_attribute: Optional[str] = None
    target_attributes: Optional[list] = None

class AttributeMappingCreate(AttributeMappingBase):
    pass

class AttributeMappingOut(AttributeMappingBase):
    id: int
    product_id: int
    model_config = {"from_attributes": True}


# --- DefaultConfiguration ---

class DefaultConfigurationBase(BaseModel):
    config_type: str
    elements: Optional[dict] = None

class DefaultConfigurationCreate(DefaultConfigurationBase):
    pass

class DefaultConfigurationOut(DefaultConfigurationBase):
    id: int
    product_id: int
    model_config = {"from_attributes": True}


# --- Intiaro Import Result ---

class IntiaroImportReport(BaseModel):
    product_id: int
    product_name: str
    sku: str
    configurations_count: int = 0
    options_count: int = 0
    variable_groups_count: int = 0
    choice_groups_count: int = 0
    predicates_count: int = 0
    events_count: int = 0
    sectional_elements_count: int = 0
    has_render_settings: bool = False
    has_product_features: bool = False
    has_menu_settings: bool = False
    attribute_mappings_count: int = 0
    default_configurations_count: int = 0
    errors: list[str] = []

File: frontend/src/types/intiaro.ts
export interface ProductFeatures {
  id: number;
  product_id: number;
  web_gltf: boolean;
  web_usdz: boolean;
  web_360: boolean;
  app_ar: boolean;
  download3d: boolean;
  com: boolean;
  has_measurements: boolean;
  resolution_x?: number;
  resolution_y?: number;
  supported_formats?: string[];
}

export interface RenderSettings {
  id: number;
  product_id: number;
  rotate_size_x?: number;
  rotate_size_y?: number;
  tile_size_x?: number;
  tile_size_y?: number;
  zoom_size_x?: number;
  zoom_size_y?: number;
  shadow_enabled: boolean;
}

export interface VariableGroup {
  id: number;
  product_id: number;
  name: string;
  slug?: string;
  icon?: string;
  icon_selected?: string;
  color?: string;
  index: number;
  priority: number;
  description?: string;
}

export interface ChoiceGroup {
  id: number;
  product_id: number;
  name: string;
  slug?: string;
  index: number;
}

export interface ProductPredicate {
  id: number;
  product_id: number;
  predicate_key: string;
  name?: string;
  type?: string;
  attribute?: string;
  operator?: string;
  compare_to?: string;
}

export interface ProductEvent {
  id: number;
  product_id: number;
  trigger_variable?: string;
  source_type?: string;
  source_variable?: string;
  destinations?: string[];
  predicate_key?: string;
  condition_attribute?: string;
  condition_operator?: string;
  condition_compare_to?: string;
}

export interface SectionalElement {
  id: number;
  product_id: number;
  element_id?: number;
  name?: string;
  display_name?: string;
  file_id?: string;
  default_variables?: Record<string, unknown>;
  includes?: Record<string, unknown>;
  positions?: Record<string, unknown>;
}

export interface MenuSettings {
  id: number;
  product_id: number;
  name?: string;
  hidden_attributes?: string[];
}

export interface AttributeMapping {
  id: number;
  product_id: number;
  source_attribute?: string;
  target_attributes?: string[];
}

export interface DefaultConfiguration {
  id: number;
  product_id: number;
  config_type: string;
  elements?: Record<string, unknown>;
}

export interface IntiaroImportReport {
  product_id: number;
  product_name: string;
  sku: string;
  configurations_count: number;
  options_count: number;
  variable_groups_count: number;
  choice_groups_count: number;
  predicates_count: number;
  events_count: number;
  sectional_elements_count: number;
  has_render_settings: boolean;
  has_product_features: boolean;
  has_menu_settings: boolean;
  attribute_mappings_count: number;
  default_configurations_count: number;
  errors: string[];
}
